### ISSUE TITLE
[ECS][Fargate]: Add EFS volumes through CloudFormation

### ISSUE THREAD

**Issue by srrengar** [aws/containers-roadmap#825]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
CloudFormation support for adding EFS volumes to ECS tasks on Fargate or EC2. 

**Which service(s) is this request for?**
Fargate, ECS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
EFS support [launched]( https://aws.amazon.com/about-aws/whats-new/2020/04/amazon-ecs-aws-fargate-support-amazon-efs-filesystems-generally-available/) for ECS on Fargate and EC2. Customers deploying through CloudFormation should also be able to use this feature.

**Comment by rpgreen** [comment]
Is there a timeline on this? It is disappointing to see feature releases without CFN support on day 1. We will have to jump through some hoops to make EFS work with our current setup.

**Comment by gary-cowell** [comment]
And I need it to work through the CDK. Which requires cloudformation first.

**Comment by guillaumesmo** [comment]
Please note you can use custom resources while a feature is not yet supported by CloudFormation

I wrote a simple proof of concept to create an ECS task definition with EFS mount through CloudFormation:
https://gist.github.com/guillaumesmo/4782e26500a3ac768888daab3c55b139

**Comment by Trandel** [comment]
:+1:

**Comment by jedis00** [comment]
Any update on this one?

**Comment by machielg** [comment]
I'm using a custom resource but the sad thing is that I need to replicate the whole task definition creation in stead of just amending the volume part. This issue is also blocking the inclusion into CDK. https://github.com/aws/aws-cdk/issues/6918

**Comment by skysb** [comment]
Hi guys, is there an update on when the cloudformation would be updated with the EFS inclusion for ECS?

**Comment by srrengar** [comment]
Hi everyone, thank you for the valuable feedback. I can assure you that this is coming soon and we will provide updates here when we have more to share. We decided to release EFS for ECS (on both EC2 and Fargate) sooner so that customers can use it and provide early feedback instead of waiting till CloudFormation was ready. We understand CloudFormation is a great way to create and update resources to use new features, and we are working on improving how we release CloudFormation support for new features going forward.

**Comment by cajames** [comment]
if anyone's still looking for a way to do this in CDK, here's a sample gist with a workaround, going down the path @guillaumesmo mentions above: https://github.com/aws/aws-cdk/issues/6240#issuecomment-635907319

The key takeaway, using CDK "Raw Overrides". Documented here: docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html

I've explained the method here: https://github.com/aws/aws-cdk/issues/6240#issuecomment-635907319

This is a temporary workaround, and once this issue is resolved, it'll be an easier switch to remove the custom task definition and add the config directly into the service. :)

**Comment by bellaxievmlyr** [comment]
Any update on how long it would take? A few days or weeks or months? I'm currently blocked by this issue. If it's a few days, I might just wait rather than spending a few days to figure out a workaround.

**Comment by machielg** [comment]
> Any update on how long it would take? A few days or weeks or months? I'm currently blocked by this issue. If it's a few days, I might just wait rather than spending a few days to figure out a workaround.

It moved from ‘in progress’ to ‘coming soon’ which also includes issues from 2019. Not a good sign 🤔

**Comment by rpgreen** [comment]
Folks, it's been over 3 months since this feature was launched and there's STILL no CFN support. This will definitely make me think twice about adopting new ECS features in the future.

**Comment by Samseppiol** [comment]
> Folks, it's been over 3 months since this feature was launched and there's STILL no CFN support. This will definitely make me think twice about adopting new ECS features in the future. CFN support is not a 'nice to have', it should be included on launch.

Are we thinking about the same company? This is AWS, cfn support isn't for atleast 12-18 months after announcing clickops.

**Comment by rpgreen** [comment]
@Samseppiol I think that's the exception more than the rule. Many features do launch with CFN support on day 1. I believe it is an internal AWS goal to do so, but the results are inconsistent at best.

**Comment by srrengar** [comment]
Hi everyone, thanks for your patience on this. CloudFormation support for EFS is in fact coming soon, and we are in the final stages of testing before we release this globally. Also, we can share that we are making changes to how we deliver CloudFormation support for ECS features. Our goal with these changes is to reduce the gap between a feature launching and that feature being supported in CloudFormation. We will continue to update this issue as more information is available.

**Comment by doogie-2** [comment]
Looking forward to this also. Trying to figure it out now

**Comment by sandeepboyapati** [comment]
eagerly waiting

**Comment by bushong1** [comment]
It's 2020.  CloudFormation has been out for 9 years.  How is this still happening?

**Comment by ghferrari** [comment]
I'm eagerly looking forward to this, too and I hope we'll be kept regularly updated on progress.

Just to give a bit of feedback to @srrengar , in my view there is simply no such thing as "releasing a feature" without cloudformation support. As far as I'm concerned,  **either** it's in cloudformation where it can be documented, placed in version control and reliably, repeatedly tested and deployed **or** it doesn't exist. As a sole developer, I don't have time for manual features and wouldn't even consider using them.

So this is me giving some feedback and a big show of encouragement to the idea of "reducing the time between a feature being released and getting cloudformation support". But I'd really like you to go all the way and understand that for many of us, a new feature is only really _released_ when it becomes available in cloudformation.

**Comment by craighurley** [comment]
All these small straws triggered my move to terraform.  So far so good, plus I get to skill up on a tool that can deploy to other cloud providers.

**Comment by gary-cowell** [comment]
> Just to give a bit of feedback to @srrengar , in my view there is simply no such thing as "releasing a feature" without cloudformation support. As far as I'm concerned, **either** it's in cloudformation where it can be documented, placed in version control and reliably, repeatedly tested and deployed **or** it doesn't exist. As a sole developer, I don't have time for manual features and wouldn't even consider using them.
> 
> So this is me giving some feedback and a big show of encouragement to the idea of "reducing the time between a feature being released and getting cloudformation support". But I'd really like you to go all the way and understand that for many of us, a new feature is only really _released_ when it becomes available in cloudformation.

Even worse, is, I have the same goals of documented code in version control and repeatable reliable testing,   **but** , I need this from the AWS  CDK.  As CDK relies on cloudformation, I have no doubt that any CDK release of this will be well behind any eventual cloudformation update.

[ I know you can mess with the cloudformation in CDK even without full support - I'd rather not do that ]

**Comment by mwarkentin** [comment]
Looks like this is supported now: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-taskdefinition.html#cfn-ecs-taskdefinition-volumes

**Comment by GeoffWilliams** [comment]
it works! I just hit the outdated lambda SDK issue where it rejects valid task definitions that use access points so now I can use this :champagne:

**Comment by frank69m** [comment]
So are you guys saying that with CDK 1.56, I can now mount EFS volumes in my task definition.... I glanced at the docs, not many good examples

**Comment by frank69m** [comment]
Ok. So I just looked at the docs. I'm looking at mounting an EFS volume to a Fargate task definition with the cdk

efs=task_efinition.add_volume(
   name="efs",
   ....
)

then within the container, I'll add the rest.

I glanced over the CDK API docs, but can't seem to find all the properties for this. Can anyone help?

**Comment by doogie-2** [comment]
Ok. so i think this is still bugged. 

I put in my fargate task def, volumes

"name": "efs-test",
"efs_volume_configuration": {
  ...
}

synth worked and after I deployed, I went to the TD and it looked like it created a  Bind Mount

**Comment by srrengar** [comment]
hey everyone, this is supported now. The official announcement will be published Monday

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-efsvolumeconfiguration.html

**Comment by ryanchapman** [comment]
Since it took me a few hours to glue CFN resources together to get this working, thought I'd post it here.  Hopefully it saves others some time.

If you run into `unsuccessful EFS utils command execution; code: 32`, then you may have security group issues.  Check the ingress on your sg associated with your mount targets as well as egress on the sg assoicated with your ECS service (`GrafanaTaskSecurityGroup` in this example).



```
AWSTemplateFormatVersion: 2010-09-09
[...]
Resources:
[...]
  GrafanaEFSVolume:
    Type: AWS::EFS::FileSystem
    Properties:
      Encrypted: true
      FileSystemTags:
        - Key: Name
          Value: !Sub "${AWS::StackName}-efs-vol"

  GrafanaEFSServerSecurityGroup:
    Type: AWS::EC2::SecurityGroup
    Properties:
      GroupName: !Sub "${AWS::StackName}-efs-server-endpoints"
      GroupDescription: Which client ip addrs are allowed to access EFS server endpoints for grafana mount
      VpcId:
        Fn::ImportValue: !Sub "${VPCStack}-VPC"
      SecurityGroupIngress:
        - IpProtocol: tcp
          FromPort: 2049
          ToPort: 2049
          CidrIp:
            Fn::ImportValue: !Sub "${VPCStack}-CIDR"
      Tags:
        - Key: Name
          Value: !Sub "${AWS::StackName}-efs-server"


  GrafanaEFSMountTarget1:
    Type: AWS::EFS::MountTarget
    Properties:
      FileSystemId: !Ref GrafanaEFSVolume
      SecurityGroups:
        - !Ref GrafanaEFSServerSecurityGroup
      SubnetId:
        Fn::Select:
          - 1
          - Fn::Split:
            - ','
            - Fn::ImportValue: !Sub "${VPCStack}-PrivateSubnets"

  GrafanaEFSMountTarget2:
    Type: AWS::EFS::MountTarget
    Properties:
      FileSystemId: !Ref GrafanaEFSVolume
      SecurityGroups:
        - !Ref GrafanaEFSServerSecurityGroup
      SubnetId:
        Fn::Select:
          - 2
          - Fn::Split:
            - ','
            - Fn::ImportValue: !Sub "${VPCStack}-PrivateSubnets"

  GrafanaTaskDefinition:
    Type: AWS::ECS::TaskDefinition
    DependsOn:
      - ECSExecutionPolicy
    Properties:
      Volumes:
        - Name: grafana
          EFSVolumeConfiguration:
            FileSystemId: !Ref GrafanaEFSVolume
            TransitEncryption: ENABLED
      ContainerDefinitions:
        - Name: grafana
          Image: !Sub "${AWS::AccountId}.dkr.ecr.${AWS::Region}.amazonaws.com/${GrafanaImage}"
          MountPoints:
            - SourceVolume: grafana
              ContainerPath: '/efs/grafana'
              ReadOnly: false
          StopTimeout: 60
          PortMappings:
            - ContainerPort: 3000
              HostPort: 3000
          LogConfiguration:
            LogDriver: awslogs
            Options:
              awslogs-region: !Ref 'AWS::Region'
              awslogs-group: !Ref ECSLogGroup
              awslogs-stream-prefix: ecs-grafana
      Cpu: '512'
      Memory: '2048'
      NetworkMode: awsvpc
      RequiresCompatibilities:
        - FARGATE
      ExecutionRoleArn: !GetAtt ECSExecutionRole.Arn
      TaskRoleArn: !GetAtt ECSTaskRole.Arn

  GrafanaTaskSecurityGroup:
    Type: AWS::EC2::SecurityGroup
    Properties:
      GroupName: !Sub "${AWS::StackName}-grafana-task"
      GroupDescription: Security group for the grafana task.
      SecurityGroupIngress:
        - IpProtocol: tcp
          FromPort: 3000
          ToPort: 3000
          SourceSecurityGroupId:
            Fn::ImportValue: !Sub "${PlatformStack}-InternalLoadBalancerServerSecurityGroup"
      SecurityGroupEgress:
        - IpProtocol: tcp
          FromPort: 2049
          ToPort: 2049
          CidrIp:
            Fn::ImportValue: !Sub "${VPCStack}-CIDR"
      VpcId:
        Fn::ImportValue: !Sub "${VPCStack}-VPC"

  GrafanaService:
    Type: AWS::ECS::Service
    Properties:
      Cluster:
        Fn::ImportValue: !Sub "${PlatformStack}-ECSCluster"
      DesiredCount: 1
      DeploymentConfiguration:
        MaximumPercent: 200
        MinimumHealthyPercent: 100
      LaunchType: FARGATE
      LoadBalancers:
        - ContainerName: grafana
          ContainerPort: 3000
          TargetGroupArn: !Ref InternalLoadBalancerTargetGroup
      NetworkConfiguration:
        AwsvpcConfiguration:
          SecurityGroups:
            - Fn::ImportValue: !Sub "${PlatformStack}-InternalLoadBalancerClientSecurityGroup"
            - !Ref GrafanaTaskSecurityGroup
          Subnets:
            Fn::Split:
              - ','
              - Fn::ImportValue: !Sub "${VPCStack}-PrivateSubnets"
      PlatformVersion: '1.4.0'
      TaskDefinition: !Ref GrafanaTaskDefinition
```

**Comment by frank69m** [comment]
Has anyone confirmed if this works with the CDK. 

Everytime I synth and try to look at my CF, it only shows

"Volumes": [
 {
   "Name": "testefs"
 }
]

And yes, I did put in the efsVolumeConfiguration

**Comment by engineal** [comment]
> Has anyone confirmed if this works with the CDK.
> 
> Everytime I synth and try to look at my CF, it only shows
> 
> "Volumes": [
> {
> "Name": "testefs"
> }
> ]
> 
> And yes, I did put in the efsVolumeConfiguration

@frank69m While I don't think this feature has been added to the CDK generated CFN resources or higher level constructs yet, I was able to use this feature through the CDK escape hatch: https://docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html

To my Fargate task definition I added a property override:
```typescript
const cfnTask = this.task.node.defaultChild as CfnTaskDefinition;
cfnTask.addPropertyOverride('Volumes', [{
    EFSVolumeConfiguration: {
        FilesystemId: this.fileSystem.fileSystemId,
        TransitEncryption: 'ENABLED'
    },
    Name: 'efs'
}]);
```

and to the container added to that task definition I added a mount point:

```typescript
container.addMountPoints({
    sourceVolume: 'efs',
    containerPath: '/data',
    readOnly: false
});
```

I have successfully used this to add an EFS mount to a Fargate task.

**Comment by cmoreno82** [comment]
I still have only the "docker_volume_configuration" option with CDK 1.56.0 and python. I use the ecs.FargateTaskDefinition.add_volume function:

 # Task Definition
          fargate_task_definition_zk = ecs.FargateTaskDefinition(
              self, "task_" + zk_name,
              family="task-" + zk_name,
              execution_role=self.role,
              memory_limit_mib=self.node.try_get_context("zookeeper_memory"),
              cpu=self.node.try_get_context("zookeeper_cpu")
          )     
           fargate_task_definition_zk.add_volume(
              name="test",
              EFSVolume....  <--- Only "docker_volume_configuration" is possible
          )
Someone else test it with cdk+python?

**Comment by doogie-2** [comment]
> > Has anyone confirmed if this works with the CDK.
> > Everytime I synth and try to look at my CF, it only shows
> > "Volumes": [
> > {
> > "Name": "testefs"
> > }
> > ]
> > And yes, I did put in the efsVolumeConfiguration
> 
> @frank69m While I don't think this feature has been added to the CDK generated CFN resources or higher level constructs yet, I was able to use this feature through the CDK escape hatch: https://docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html
> 
> To my Fargate task definition I added a property override:
> 
> ```ts
> const cfnTask = this.task.node.defaultChild as CfnTaskDefinition;
> cfnTask.addPropertyOverride('Volumes', [{
>     EFSVolumeConfiguration: {
>         FilesystemId: this.fileSystem.fileSystemId,
>         TransitEncryption: 'ENABLED'
>     },
>     Name: 'efs'
> }]);
> ```
> 
> and to the container added to that task definition I added a mount point:
> 
> ```ts
> container.addMountPoints({
>     sourceVolume: 'efs',
>     containerPath: '/data',
>     readOnly: false
> });
> ```
> 
> I have successfully used this to add an EFS mount to a Fargate task.

OK. Thanks @engineal I have successfully used the escape hatch in CDK to make it work! Finally! Thanks for your help.

**Comment by srrengar** [comment]
Here is the official announcement - https://aws.amazon.com/about-aws/whats-new/2020/08/amazon-ecs-announces-cloudformation-support-for-amazon-efs-volumes/

**Comment by frank69m** [comment]
great. works now with CDK and the @engineal solution above.

**Comment by paveljos** [comment]
Just to close the gap for anyone looking for a python solution based on @engineal 's submission, here's what I implemented in CDK:
```
cfn_task = self.task_definition.node.default_child
cfn_task.add_property_override("Volumes", [{
   "EFSVolumeConfiguration": {
      "FilesystemId": efs_id.file_system_id,
      "TransitEncryption": "ENABLED"
   },
   "Name": "efs"
}])
```

**Comment by skindc** [comment]
For anyone still having issues with this, I seems ensuring that you explicitly set the platform version in your containers definition to '1.4.0' solves the problem. When using 'latest' this is actually implying 1.3.0 which seems does not support the latest announcements.

**Comment by yogeshjoshi** [comment]
> For anyone still having issues with this, I seems ensuring that you explicitly set the platform version in your containers definition to '1.4.0' solves the problem. When using 'latest' this is actually implying 1.3.0 which seems does not support the latest announcements.

Yes @skindc , yes this has been solved in version 1.4.0 in August this year.
Latest version in definition refers 1.3.0

**Comment by mreferre** [comment]
@skindc / @yogeshjoshi tangential to this thread, we have an open issue around how to better manage/communicate new Platform Versions. If you have opinions or feedback please leave a comment at https://github.com/aws/containers-roadmap/issues/1069 . Thanks!

**Comment by eob** [comment]
@engineal I'm using your solution above, but when using `ecs_patterns.ApplicationLoadBalancedFargateService` with the `taskImageOptions` I can't figure out how to access my `container` object, as in:

```
container.addMountPoints({
    sourceVolume: 'efs',
    containerPath: '/data',
    readOnly: false
});
```

Do you know if it's still possible to attach an EFS volume to a container that has been created with:

```
      taskImageOptions: { 
        image: ContainerImage.fromEcrRepository(...),
        environment: { ... },
        containerPort: 8080,
        enableLogging: true,
      },
```

**Comment by engineal** [comment]
Hi @eob and anyone else who is interested. I believe the property override I had suggested before isn't required anymore, as these props are now supported by the aws-cdk constructs.

I can replace my previous use of the property override with a call to the `addVolume` method on my task definition:
```typescript
this.task.addVolume({
    name: "efs",
    efsVolumeConfiguration: {
        fileSystemId: this.fileSystem.fileSystemId,
        transitEncryption: "ENABLED"
    }
});
```

And just like before, to the container added to that task definition I added a mount point referencing that volume:
```typescript
container.addMountPoints({
    sourceVolume: 'efs',
    containerPath: '/data',
    readOnly: false
});
```

Now for your example @eob, if you've created your service using the ecs-patterns `ApplicationLoadBalancedFargateService` construct like so:
```typescript
const service = new ApplicationLoadBalancedFargateService(this, 'service', {
    taskImageOptions: {
        image: ContainerImage.fromEcrRepository(...),
        environment: { ... },
        containerPort: 8080,
        enableLogging: true,
    }
});
```

you can access the task definition created by that service using `service.taskDefinition`, and add the EFS volume to the task definition like so:

```typescript
service.taskDefinition.addVolume({
    name: "efs",
    efsVolumeConfiguration: {
        fileSystemId: this.fileSystem.fileSystemId,
        transitEncryption: "ENABLED"
    }
});
```

Now the `ApplicationLoadBalancedFargateService` construct adds a single container to the task definition, which will either be named 'web' or the value you assign `containerName` in your `taskImageOptions`. This means you can access that container by using `taskDefinition.findContainer('web')` (or `taskDefinition.findContainer(containerName)` if you specified  `containerName` in your `taskImageOptions`), Additionally, since the container added by the `ApplicationLoadBalancedFargateService` is the first container added, you can access it by using `taskDefinition.defaultContainer` as well.

So, to add the mount point to the container created by the `ApplicationLoadBalancedFargateService`, you can do:
```typescript
service.taskDefinition.findContainer('web')?.addMountPoints({
    sourceVolume: 'efs',
    containerPath: '/data',
    readOnly: false
});
```
or 
```typescript
service.taskDefinition.defaultContainer?.addMountPoints({
    sourceVolume: 'efs',
    containerPath: '/data',
    readOnly: false
});
```

**Comment by eob** [comment]
@engineal thank you for so kindly explaining that!

Your comment got everything deploying properly, but I'm still getting permissions errors when I try to actually access the volume.

I've seen a few articles about configuring AccessPoints with EFS, but most Fargate tutorials proceed without them as if they're not actually necessary. Do you know if there any permissions gotchas that I might not be noticing?

This is what my code looks like -- I copy-paste-customized the ApplicationLoadBalancedFargateService code so that it's easier to just pass in mounts as an option.

```
// FILESYSTEM
const fs = new efs.FileSystem(this, "IndexFilesystem", {
  vpc: vpc,
  performanceMode: efs.PerformanceMode.GENERAL_PURPOSE,
  throughputMode: efs.ThroughputMode.BURSTING,
  fileSystemName: "indexFilesystem"
})
  
// VOLUME
const volume = {      
  name: "indexVolume",
  efsVolumeConfiguration: {
    fileSystemId: fs.fileSystemId,
    TransitEncryption: 'ENABLED',
  },
};

// this is a FargateTaskDefinition object
this.taskDefinition.addVolume(volume)  


// MOUNT POINT
const mountPoint = {
  containerPath: "/data",
  sourceVolume: volume.name,
  readOnly: false,
};

// This is the container created inside ApplicationLoadBalancedFargateService
container.addMountPoints(mountPoint)

// PROVIDE ACCESS
// this.service is the FargateService
this.service.connections.allowFrom(fs, Port.tcp(2049));
this.service.connections.allowTo(fs, Port.tcp(2049));
```

**Comment by engineal** [comment]
> 
> 
> @engineal thank you for so kindly explaining that!
> 
> Your comment got everything deploying properly, but I'm still getting permissions errors when I try to actually access the volume.
> 
> I've seen a few articles about configuring AccessPoints with EFS, but most Fargate tutorials proceed without them as if they're not actually necessary. Do you know if there any permissions gotchas that I might not be noticing?
> 
> This is what my code looks like -- I copy-paste-customized the ApplicationLoadBalancedFargateService code so that it's easier to just pass in mounts as an option.
>

@eob Would you happen to be able to describe or include in a comment the error message that you receive?

**Comment by sukrit007** [comment]
@eob  What is the user (UID/GID) of the user running within the container?   You might have to create an access point for specific Posix user to get around Posix permissions.  https://docs.aws.amazon.com/efs/latest/ug/accessing-fs-nfs-permissions.html
